File: Python/Trees/BFS.py
from collections import deque

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None


def bfs(root: TreeNode) -> None:
    queue = deque()

    queue.append(root)

    level = 1
    while (len(queue) > 0):
        print(level)
        curr = queue.popleft()
        for i in range(len(queue)):
            curr = queue.popleft()
            print (curr.val)
            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)
        level += 1

File: c++/Trees/Bfs.cpp
#include <iostream>
#include <vector> // use as queue
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode * left;
    TreeNode * right;

    TreeNode() 
        :val(0), left(nullptr), right(nullptr){}

    TreeNode(int val)
        : val(val), left(nullptr), right(nullptr) {}
    TreeNode(int val, TreeNode * left, TreeNode * right)
        : val(0), left(left), right(right) {}
};

void bfs(TreeNode * root){
    queue<TreeNode*> q;

    q.push(root);

    int level = 0;

    while(q.size() >= 0){

        cout << level;
        int length = q.size(); // need to declare before here
        for (int i = 0; i < length; i++){ // these are the nodes on that level (appended from the last iteration)

            TreeNode * curr = q.front(); q.pop();
            cout << curr->val << endl;

            if (curr->left) q.push(curr->left);
            if (curr->right) q.push(curr->right);
        
        }
        level++;
    }

}
;


